File: front-end/lib/services/income_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

import '../core/constants/api_endpoints.dart';
import 'auth_service.dart';

/// Tier C: logs manual income events (no bank integration); payday prediction is server-side via GET /api/insights/forecast's `expectedIncome`, not fetched here.
class IncomeService {
  final _authService = AuthService();

  Future<Map<String, String>> _authHeaders() async {
    final token = await _authService.getToken();
    return {
      'Content-Type': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  /// POST /api/income
  Future<Map<String, dynamic>> logIncome({
    required double amount,
    String? source,
    DateTime? receivedDate,
  }) async {
    try {
      final response = await http.post(
        Uri.parse(ApiEndpoints.income),
        headers: await _authHeaders(),
        body: jsonEncode({
          'amount': amount,
          if (source != null && source.isNotEmpty) 'source': source,
          // receivedDate is a DATE-only column server-side — trim the time
          // component toIso8601String() would otherwise include.
          if (receivedDate != null)
            'receivedDate': receivedDate.toIso8601String().substring(0, 10),
        }),
      );

      final body = jsonDecode(response.body) as Map<String, dynamic>;
      if (response.statusCode == 201) {
        return body;
      } else {
        throw Exception(body['message'] ?? 'Failed to log income');
      }
    } catch (e) {
      throw Exception('Error logging income: $e');
    }
  }

  /// GET /api/income — last 6 months of logged income.
  Future<List<dynamic>> fetchIncomeHistory() async {
    try {
      final response = await http.get(
        Uri.parse(ApiEndpoints.income),
        headers: await _authHeaders(),
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body) as List<dynamic>;
      } else {
        throw Exception(
          'Failed to fetch income history (status ${response.statusCode}): ${response.body}',
        );
      }
    } catch (e) {
      throw Exception('Error fetching income history: $e');
    }
  }

  /// PUT /api/income/:id
  Future<Map<String, dynamic>> updateIncome(
    String incomeId, {
    double? amount,
    String? source,
    DateTime? receivedDate,
  }) async {
    try {
      final response = await http.put(
        Uri.parse(ApiEndpoints.incomeUpdate(incomeId)),
        headers: await _authHeaders(),
        body: jsonEncode({
          'amount': ?amount,
          'source': ?source,
          if (receivedDate != null)
            'receivedDate': receivedDate.toIso8601String().substring(0, 10),
        }),
      );

      final body = jsonDecode(response.body) as Map<String, dynamic>;
      if (response.statusCode == 200) {
        return body;
      } else {
        throw Exception(body['message'] ?? 'Failed to update income entry');
      }
    } catch (e) {
      throw Exception('Error updating income entry: $e');
    }
  }

  /// DELETE /api/income/:id
  Future<void> deleteIncome(String incomeId) async {
    try {
      final response = await http.delete(
        Uri.parse(ApiEndpoints.incomeDelete(incomeId)),
        headers: await _authHeaders(),
      );

      if (response.statusCode != 200) {
        final body = jsonDecode(response.body) as Map<String, dynamic>;
        throw Exception(body['message'] ?? 'Failed to delete income entry');
      }
    } catch (e) {
      throw Exception('Error deleting income entry: $e');
    }
  }
}

File: back-end/src/services/income.service.js
const incomeModel = require('../models/income.model');

// POST /api/income — logs a single income event; manual entry since there's no bank integration to auto-detect it.
exports.logIncome = async (req, res) => {
  const { amount, source, receivedDate } = req.body;

  const parsedAmount = Number(amount);
  if (!Number.isFinite(parsedAmount) || parsedAmount <= 0) {
    return res.status(400).json({ message: 'amount must be a positive number' });
  }

  try {
    const result = await incomeModel.createIncome({
      userId: req.user.userId,
      amount: parsedAmount,
      source,
      receivedDate,
    });
    return res.status(201).json({ message: 'Income logged successfully', data: result.rows[0] });
  } catch (err) {
    console.error('Log income error', err);
    return res.status(500).json({ message: 'Failed to log income', error: err.message });
  }
};

// GET /api/income — last 6 months of logged income, for display/editing.
exports.listIncome = async (req, res) => {
  try {
    const result = await incomeModel.listIncomeForUser(req.user.userId);
    return res.status(200).json(result.rows);
  } catch (err) {
    console.error('List income error', err);
    return res.status(500).json({ message: 'Failed to fetch income', error: err.message });
  }
};

exports.updateIncome = async (req, res) => {
  const { id } = req.params;
  const { amount, source, receivedDate } = req.body;

  if (amount !== undefined && (!Number.isFinite(Number(amount)) || Number(amount) <= 0)) {
    return res.status(400).json({ message: 'amount must be a positive number' });
  }

  try {
    const result = await incomeModel.updateIncome(id, req.user.userId, {
      amount: amount !== undefined ? Number(amount) : undefined,
      source,
      receivedDate,
    });
    if (result.rows.length === 0) {
      return res.status(404).json({ message: 'Income entry not found' });
    }
    return res.status(200).json({ message: 'Income entry updated successfully', data: result.rows[0] });
  } catch (err) {
    console.error('Update income error', err);
    return res.status(500).json({ message: 'Failed to update income entry', error: err.message });
  }
};

exports.deleteIncome = async (req, res) => {
  const { id } = req.params;

  try {
    const result = await incomeModel.deleteIncome(id, req.user.userId);
    if (result.rowCount === 0) {
      return res.status(404).json({ message: 'Income entry not found' });
    }
    return res.status(200).json({ message: 'Income entry deleted successfully' });
  } catch (err) {
    console.error('Delete income error', err);
    return res.status(500).json({ message: 'Failed to delete income entry', error: err.message });
  }
};
